platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -326,29 +330,39 @@ We will pass the full round info (metadata + events) to all listeners, and they
 and pass that roster to the consensus system. **This must be deterministic**. Each node in the network must assign the
 same roster to the same rounds. However, Consensus and Execution run in completely different threads, and possibly,
 in different processes.
 
-With this architecture, we introduce a new concept into `EventCore`. `EventCore` has a `payload`, which is a `byte[]`
-that the Execution layer can use to carry whatever payload is meaningful to it. The consensus layer needs an analogous
-section of the `EventCore` into which messages particular to Consensus can be placed. This section will be called
-`repeated control_messages`. A `ControlMessage` is a protobuf message used to control, in some way, the consensus
-system.
-
-TODO: I'm seriously thinking of having "system_transactions" and "user_transactions" instead of "control_messages" and
-"payload".
-
-`ControlMessage` has a one-of, among which is `RosterChangeMessage`. This message will have the new roster, and be
-signed by the network as a whole. Each honest node will submit a `RosterChangeMessage`, and the Hashgraph module will
-recognize these messages. When it encounters the **first** such message for a given roster in a round that has come to
-consensus, the very next round will use that new activated roster.
+Each `EventCore` will have the ability to carry two different types of transactions: `user_transactions` and
+`system_transactions`. A "system" transaction is a transaction known by Consensus, but not Execution, while "user"
+transactions are those supplied by Execution, and unknown to Consensus. By separating these two fields, we make it very
+inexpensive for logic in Consensus to iterate over and handle "system" transactions regardless of the number of user
+transactions in the event.
+
+Each node, individually, decides deterministically at some round that it is time to adopt a new roster, and
+deterministically select the exact same roster. For example, this might happen on the first transaction after midnight,
+or it might happen every 30 minutes, or it might happen after TSS keys have been generated and messages exchanged with
+peers. However it happens, every node will deterministically decide on the same exact roster at the same consensus-time
+moment and call their respective Consensus systems to `changeRoster`. This call is then handled by the Event Intake
+system. Each node's Event Intake will take this roster and include it in the next event to be gossiped as a "system"
+transaction.
+
+Each node's Hashgraph module will watch for events containing new rosters. If it sees a new roster, it will hash it,
+and keep track in an in-memory data structure of how many nodes (and their consensus weight) have suggested this same
+new roster. When 2/3 of nodes by stake weight have suggested the same roster, it becomes effective. The Hashgraph
+module includes the current roster hash and the full roster in the metadata for every round. In this way, by simply
```

Comment:
I don't see why it is necessary to include the full roster in the round metadata? What is that needed for?